### PR TITLE
JTJ-28 Login Authentication 

### PR DESCRIPTION
**Pull request by mla-wtag**
[JTJ-28](https://welltravel.atlassian.net/jira/software/projects/JTJ/boards/44?selectedIssue=JTJ-28)
[JTJ-30](https://welltravel.atlassian.net/jira/software/projects/JTJ/boards/44?selectedIssue=JTJ-30)

**Description**
- The function for the profile pic is not yet implemented.
- The validations do not have a message yet.
- I apologize for the large Pull Request, but I wanted to implement RSpec and internationalization.
- The video is too large for direct viewing; you will need to download and unzip it.
- In the future, the PRs, videos will be smaller and only one User story. 

[view_of_the_App.zip](https://github.com/mla-wtag/ml_Journey/files/12583345/view_of_the_App.zip)

### CODE REVIEW

File: .rufo [hunk 1]
```diff
@@ -0,0 +1 @@
+quote_style :single
```
Comment (evk-wtag) [inline .rufo]: Let's add a new line at the end of files

File: app/views/users/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<%= link_to t('back'), user_path(@user) %>
+<h1>Edit Profile</h1>
```
Comment (evk-wtag) [inline app/views/users/edit.html.erb]: Missing i18n translation

File: app/views/users/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,95 @@
+<%= link_to t('back'), root_path %>
+<h1>Create User</h1>
```
Comment (evk-wtag) [inline app/views/users/new.html.erb]: Missing i18n translation

File: config/locales/en.yml [hunk 4]
```diff
@@ -1,33 +1,27 @@
...
+  task:   'Tasks'
+  members: 'Team Members'
+  back: 'Back'
+
+  buttons:
+  log_out: 'Log Out'
+  create_user: 'Create Profile'
+  register: 'Register'
+  edit: 'Edit Profile'
+  delete: 'Delete'
```
Comment (evk-wtag) [inline config/locales/en.yml]: New line at EOF please

File: db/migrate/20230823111409_create_users.rb [hunk 5]
```diff
@@ -9,6 +9,8 @@ def up
       t.string :designation
       t.binary :profile_picture
       t.string :email
+      t.string :password_digest
```
Comment (evk-wtag) [inline db/migrate/20230823111409_create_users.rb]: We won't edit the existing migrations!
 Let's add a new migration to add this field 😅

File: db/schema.rb [hunk 6]
```diff
@@ -12,38 +12,38 @@
 
 ActiveRecord::Schema[7.0].define(version: 2023_08_23_124431) do
   # These are extensions that must be enabled in order to support this database
-  enable_extension "plpgsql"
+  enable_extension 'plpgsql'
```
Comment (evk-wtag) [inline db/schema.rb]: Did you manually update this file? 🤔

Comment (mla-wtag) [reply db/schema.rb]: Yes, I went over the file with the formatter and changed the quotation marks

File: app/models/user.rb [hunk 7]
```diff
@@ -1,2 +1,11 @@
...
+class User < ApplicationRecord
+  has_secure_password
+  validates :firstname, presence: true, uniqueness: true
+  validates :lastname, presence: true
+  validates :employee_id, presence: true
+  validates :date_of_birth, presence: true
+  validates :joining_day, presence: true
+  validates :designation, presence: true
+  validates :email, presence: true
+  validates :password, presence: true
```
Comment (chm-wtag) [inline app/models/user.rb]: `validates_presence_of :lastname, :employee_id, :date_of_birth ...`

File: app/views/users/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    <%= f.label t('profile_picture') %><br>
+    <%= f.text_field :profile_picture %><br>
+  </div>
+  <div>
+    <%= f.label t('Email') %><br>
+    <%= f.text_field :email %><br>
+  </div>
+  <div>
+    <%= f.label t('password') %><br>
+    <%= f.text_field :password %><br>
```
Comment (chm-wtag) [inline app/views/users/edit.html.erb]: can we use hidden field here?

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+<%= link_to t('edit'), edit_user_path %>
+<br><br><br>
+<%= link_to t('delete'), user_path(@user), data: {
+                                             turbo_method: :delete,
+                                             turbo_confirm: 'Are you sure?',
+                                           } %>
+<br><br>
+<table>
+    <tr>
+      <th>profile Picture</th>
```
Comment (evk-wtag) [inline app/views/users/show.html.erb]: Let's use i18n for every hardcoded texts

File: config/locales/en.yml [hunk 10]
```diff
@@ -1,33 +1,28 @@
...
-# true, false, on, off, yes, no
-#
-# Instead, surround them with single quotes.
-#
-# en:
-#   "true": "foo"
-#
-# To learn more, please read the Rails Internationalization guide
-# available at https://guides.rubyonrails.org/i18n.html.
+en:
```
Comment (chm-wtag) [inline config/locales/en.yml]: use i18n-tasks gem and run ` i18n-tasks health`

File: config/locales/en.yml [hunk 11]
```diff
@@ -1,33 +1,28 @@
...
-# Instead, surround them with single quotes.
-#
-# en:
-#   "true": "foo"
-#
-# To learn more, please read the Rails Internationalization guide
-# available at https://guides.rubyonrails.org/i18n.html.
+en:  
+  attributes: 
+  firstname: 'Firstname'
```
Comment (evk-wtag) [inline config/locales/en.yml]: Shouldn't it be under the attributes hash?

We don't need to put any quotations for the strings in this file unless we need to display a quote in text.

You may want to check the i18n activerecord key that allows you to specify a model name and raise errors accordingly. Check this out: https://guides.rubyonrails.org/i18n.html#translations-for-active-record-models

File: spec/factories/users.rb [hunk 12]
```diff
@@ -0,0 +1,12 @@
+FactoryBot.define do
+  factory :user do
+    firstname { 'Michael' }
+    lastname { 'Lavelanet' }
+    employee_id { '12345' }
+    date_of_birth { '2004-12-12'.to_date }
+    joining_day { '2004-12-12'.to_date }
+    designation { 'Junior Software Engineer' }
+    email { 'Mail' }
+    password { 'code' }
```
Comment (chm-wtag) [inline spec/factories/users.rb]: use FFaker gem for generating random data

File: spec/models/users_spec.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
```
Comment (chm-wtag) [inline spec/models/users_spec.rb:1]: study about shoulda-matcher gem and use that

File: spec/factories/users.rb [hunk 14]
```diff
@@ -0,0 +1,12 @@
+FactoryBot.define do
+  factory :user do
+    firstname { 'Michael' }
+    lastname { 'Lavelanet' }
+    employee_id { '12345' }
+    date_of_birth { '2004-12-12'.to_date }
+    joining_day { '2004-12-12'.to_date }
+    designation { 'Junior Software Engineer' }
+    email { 'Mail' }
```
Comment (evk-wtag) [inline spec/factories/users.rb]: Let's add the gem Ffaker to generate random texts and numbers

File: spec/controllers/users_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+      get :show, params: { id: 1 }
+
+      expect(assigns :user).to eq user1
+    end
+
+    it 'renders the :show template' do
+      get :show, params: { id: 1 }
+      expect(response).to render_template :show
+    end
+  end
```
Comment (chm-wtag) [inline spec/controllers/users_controller_spec.rb]:
```
  describe 'GET user#show' do
    get :show, params: { id: 1 }
    it 'displays the current user attributes with variables' do
      expect(assigns :user).to eq user1
    end

    it 'renders the :show template' do
      expect(response).to render_template :show
    end
  end
```
common line can be put in the common block

File: app/models/user.rb [hunk 16]
```diff
@@ -1,2 +1,11 @@
-class User < ApplicationRecord 
+class User < ApplicationRecord
+  has_secure_password
+  validates :firstname, presence: true, uniqueness: true
+  validates :lastname, presence: true
+  validates :employee_id, presence: true
+  validates :date_of_birth, presence: true
+  validates :joining_day, presence: true
+  validates :designation, presence: true
+  validates :email, presence: true
```
Comment (chm-wtag) [inline app/models/user.rb]: add regex for email validation

File: spec/controllers/users_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+        delete :destroy, params: { id: 1 }
+      }.to change(User, :count).by(-1)
+    end
+    it 'redirects to the root path' do
+      delete :destroy, params: { id: 1 }
+      expect(response).to redirect_to(root_path)
+    end
+  end
+
+  # describe 'DELETE #destroy' do
```
Comment (tgi-wtag) [inline spec/controllers/users_controller_spec.rb]: delete comments if not needed

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -1,2 +1,53 @@
 class UsersController < ApplicationController
```
Comment (lra-wtag) [inline app/controllers/users_controller.rb:1]: use this -> load_and_authorize_resource at the top of the controller then you wont need to create instances anymore

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -1,2 +1,53 @@
...
+    @user = User.new(user_params)
+    if @user.save
+      redirect_to root_path
+    else
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def show
+    @user = User.find(params[:id])
```
Comment (lra-wtag) [inline app/controllers/users_controller.rb]: after introducing load_and_authorize_resource you can write it as:
`def show; end`

File: app/models/ability.rb [hunk 20]
```diff
@@ -1,32 +0,0 @@
-# frozen_string_literal: true
```
Comment (lra-wtag) [inline app/models/ability.rb:1]: we need this ability file

File: config/locales/en.yml [hunk 21]
```diff
@@ -1,33 +1,28 @@
...
+  members: 'Team Members'
+  back: 'Back'
+
+  buttons:
+  log_out: 'Log Out'
+  create_user: 'Create Profile'
+  register: 'Register'
+  edit: 'Edit Profile'
+  delete: 'Delete'
+
```
Comment (lra-wtag) [inline config/locales/en.yml]: add an extra line at the end of the file please

File: app/helpers/application_helper.rb [hunk 22]
```diff
@@ -1,2 +0,0 @@
-module ApplicationHelper
-end
```
Comment (lra-wtag) [inline app/helpers/application_helper.rb]: we usually keep this file

File: app/views/users/new.html.erb [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+        <%= @user.errors.full_messages_for(:firstname).join(', ') %>
+      </div><br>
+    <% end %>
+  </div>
+  <div>
+    <%= f.label t('profile_picture') %><br>
+    <%= f.text_field :profile_picture %><br>
+  </div>
+  <div>
+    <%= f.label t('Email') %><br>
```
Comment (lra-wtag) [inline app/views/users/new.html.erb]:
```suggestion
    <%= f.label t('email') %><br>
```

**Review by lra-wtag — CHANGES_REQUESTED** [review]

**Review by chm-wtag — APPROVED** [review]
Good work 🎉

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -1,2 +1,47 @@
 class UsersController < ApplicationController
+  load_and_authorize_resource
+
+  def new
+  end
```
Comment (evk-wtag) [inline app/controllers/users_controller.rb]: Is this needed?

File: app/controllers/users_controller.rb [hunk 25]
```diff
@@ -1,2 +1,47 @@
...
+    else
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def show
+  end
+
+  def edit
+  end
```
Comment (evk-wtag) [inline app/controllers/users_controller.rb]: Let's remove if not needed

File: app/models/user.rb [hunk 26]
```diff
@@ -1,2 +1,5 @@
-class User < ApplicationRecord 
+class User < ApplicationRecord
+  has_secure_password
+  validates_presence_of :firstname, :lastname, :employee_id, :date_of_birth, :joining_day, :designation, :password
+  validates :email, presence: true, format: { with: /\A[^@\s]+@[^@\s]+\z/, message: 'must be a valid email address' }
```
Comment (evk-wtag) [inline app/models/user.rb]: Can we use i18n for the text?

**Review by evk-wtag — APPROVED** [review]

File: db/migrate/20230913111312_alter_users.rb [hunk 27]
```diff
@@ -0,0 +1,9 @@
+class AlterUsers < ActiveRecord::Migration[7.0]
+  def up
+    add_column('users', 'password_digest', :string)
```
Comment (evk-wtag) [inline db/migrate/20230913111312_alter_users.rb]:
```suggestion
    add_column :users, :password_digest, :string
```

File: db/migrate/20230913111312_alter_users.rb [hunk 28]
```diff
@@ -0,0 +1,9 @@
...
+  end
+
+  def down
+    remove_column('users', 'password_digest')
```
Comment (evk-wtag) [inline db/migrate/20230913111312_alter_users.rb]:
```suggestion
    remove_column :users, :password_digest
```

File: app/models/ability.rb [hunk 29]
```diff
@@ -1,32 +1,9 @@
...
-    # The third argument is an optional hash of conditions to further filter the
-    # objects.
-    # For example, here the user can only update published articles.
-    #
-    #   can :update, Article, published: true
-    #
-    # See the wiki for details:
-    # https://github.com/CanCanCommunity/cancancan/blob/develop/docs/define_check_abilities.md
+    user ||= User.new
+    can [:create, :read, :update, :destroy], User
```
Comment (lra-wtag) [inline app/models/ability.rb]: can %i(:create, :read, :update, :destroy), User

File: app/views/users/show.html.erb [hunk 30]
```diff
@@ -0,0 +1,46 @@
+<h1> <%= t('views.welcome') + ' ' + current_user.firstname %></h1>
```
Comment (lra-wtag) [inline app/views/users/show.html.erb]: this additional space can be added with the translations itself

File: config/routes.rb [hunk 31]
```diff
@@ -1,2 +1,10 @@
 Rails.application.routes.draw do
+  root 'user_sessions#new'
+  resources :user_sessions, only: [:new, :create, :destroy]
+  resources :users
+  resources :users do
+    member do
+      get :delete # adding the delete action manual, because rails doesn't give it to you.
```
Comment (lra-wtag) [inline config/routes.rb]: why do we need this delete function ? rails already has the destroy function by default right ?

File: db/migrate/20230913111312_alter_users.rb (hunk 27 above)
Comment (lra-wtag) [inline db/migrate/20230913111312_alter_users.rb]:
```suggestion
    add_column :users, :password_digest, :string
```

File: db/migrate/20230913111312_alter_users.rb (hunk 28 above)
Comment (lra-wtag) [reply db/migrate/20230913111312_alter_users.rb]: please address this

**Review by lra-wtag — CHANGES_REQUESTED** [review]

File: config/locales/en.yml [hunk 32]
```diff
@@ -1,33 +1,48 @@
...
+    create_user: Create Profile
+    edit: Edit Profile
+    delete: Delete
+    log_out: Log Out
+  views:
+    journal: Log Journal
+    login: Login
+    members: Team Members
+    profile: Profile
+    welcome: Welcome, %{name}
```
Comment (lra-wtag) [inline config/locales/en.yml]:
```suggestion
    welcome: 'Welcome, %{name}'
```

File: config/routes.rb [hunk 33]
```diff
@@ -1,2 +1,6 @@
 Rails.application.routes.draw do
+  root 'user_sessions#new'
+  resources :user_sessions, only: [:new, :create, :destroy]
+  resources :users
+  resources :users
```
Comment (lra-wtag) [inline config/routes.rb]: why two "resources :users" ?

**Review by lra-wtag — CHANGES_REQUESTED** [review]

**Review by lra-wtag — APPROVED** [review]

**Pull request merged**